File: app/src/main/java/com/xjm/xxd/dribbird/bus/RxBus.java
package com.xjm.xxd.dribbird.bus;


import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;

/**
 * Created by queda on 2016/11/21.
 */

public class RxBus{

    private static volatile RxBus mInstance;

    private final FlowableProcessor<IBusEvent> mActual;

    private RxBus() {
        mActual = PublishProcessor.<IBusEvent>create().toSerialized();
    }

    public static RxBus getInstance() {
        if (mInstance == null) {
            synchronized(RxBus.class) {
                if (mInstance == null) {
                    mInstance = new RxBus();
                }
            }
        }
        return mInstance;
    }

    public void postEvent(IBusEvent event) {
        mActual.onNext(event);
    }

    public <T> Flowable<T> register(Class<T> clazz) {
        return mActual.ofType(clazz);
    }

}

File: skeleton/src/main/java/com/xjm/xxd/skeleton/util/rx/RxBus.kt
package com.xjm.xxd.skeleton.util.rx

import com.jakewharton.rxrelay2.PublishRelay
import io.reactivex.Observable
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by queda on 2018/3/14.
 */

class RxBus {

    private val mBus = PublishRelay.create<Any>().toSerialized()

    private val mStickyMap = ConcurrentHashMap<Class<out Any>, Any>()

    public fun post(event: Any) {
        mBus.accept(event)
    }

    public fun <T> receive(eventType: Class<T>): Observable<T> {
        return mBus.ofType(eventType)
    }

    public fun postSticky(stickyEvent: Any) {
        mBus.accept(stickyEvent)
        mStickyMap[stickyEvent::class.java] = stickyEvent
    }

    public fun <T> receiveSticky(eventType: Class<T>): Observable<T> {
        val stickyEvent = mStickyMap[eventType]
        val receiveObservable = receive(eventType)
        return if (stickyEvent == null) {
            receiveObservable
        } else {
            Observable.create<T> { emitter ->
                if (!emitter.isDisposed) {
                    emitter.onNext(stickyEvent as T)
                }
            }.mergeWith(receiveObservable)
        }
    }

    public fun clearSticky() {
        mStickyMap.clear()
    }

    companion object {
        @JvmStatic
        val instance by lazy { HOLDER.INSTANCE }
    }

    private object HOLDER {
        val INSTANCE = RxBus()
    }

}